cloud-functions/cf-dsm-file-scanner/src/main/java/org/broadinstitute/ddp/cf/DSMSomaticFileScanner.java

```diff
@@ -0,0 +1,334 @@
+        }
+
+        if (fileName == null || fileName.isBlank()) {
+            var logMessage = eventLogPrefix + ": "
+                    + "Bucket filename is missing in message (attribute " + ATTR_OBJECT_ID + ")";
+            log.atSevere().log(logMessage);
+            throw new RuntimeException(logMessage);
+        }
+
+        log.atInfo().log("Looking up file: bucket=%s name=%s", bucketName, fileName);
+        Blob blob = storage.get(bucketName, fileName);
+        if (blob == null || !blob.exists()) {
+            log.atWarning().log("Could not find file: bucket=%s name=%s", bucketName, fileName);
+            // This might be a duplicate (due to pubsub at-least-once delivery), in which case the
+            // file might have already been moved. Exit gracefully so we don't incur a cold start.
+            return;
+        }
+
+        BlobId blobId = blob.getBlobId();
+        Instant createdAt = Instant.ofEpochMilli(blob.getCreateTime());
+        log.atInfo().log("Found file that was created at %s", createdAt.toString());
+
+        log.atInfo().log("Scanning file: %s", blobId.toString());
+
+        var clamav = new Client(clamdAddress.getHostName(), clamdAddress.getPort());
+
+        ScanResult result;
+        try {
+            if (!clamav.ping()) {
+                var logMessage = eventLogPrefix + ": "
+                        + "clamd host " + clamdAddress + " did not respond to a PING";
+                log.atSevere().log(logMessage);
+                throw new RuntimeException(logMessage);
+            }
+
+            log.atFine().log("%s: clamd host %s responded to a PING", eventLogPrefix, clamdAddress.toString());
+
+            var bucketDataStream = Channels.newInputStream(blob.reader());
+            result = runClamscan(clamav, bucketDataStream);
+
+            switch (result) {
+                case INFECTED:
+                    log.atInfo().log("%s: malware was identified in object %s/%s",
+                            eventLogPrefix, blob.getBucket(), blob.getName());
+                    deleteFileFromBucket(blob.getBucket(), blob.getName(), gcpProjectId);
+                    break;
+
+                case CLEAN:
+                    log.atInfo().log("%s: no malware identified in object %s/%s",
+                            eventLogPrefix, blob.getBucket(), blob.getName());
+                    moveFileIntoFinalBucket(blob.getBucket(), blob.getName(), gcpProjectId, finalBucket);
+                    break;
+                default:
+                    throw new RuntimeException("unreachable");
+            }
+        } catch (IOException ioe) {
+            throw new RuntimeException("connection with " + clamdAddress + " was unexpectedly closed", ioe);
+        }
+    }
+
+    private void publishAntivirusResultMessageToDSM(AntivirusMessage antivirusMessage) {
+        if (publisher.isPresent()) {
+            ByteString data = ByteString.copyFrom(antivirusMessage.toString(), StandardCharsets.UTF_8);
```

Comment:
How about converting to JSON so we send a simple object with just:
{
"bucketId":"<original bucket>",
"objectId":"<original object id>",
"newBucketId":"<new bucket>",
"newObjectId"":"<new object ID>",
"scanResult":"<scan result>"
}

Reply:
Implemented